### DISCUSSION TITLE
cli11 check never called after transform validator

### DISCUSSION THREAD

**Discussion by dhsu123** [CLIUtils/CLI11#675] [Q&A] [answered]
Using CLI11 when i call a transform mutating validator followed by a call to the check validator, check never gets called. If a non mutating transform is called before check, everything is fine and check is called. for example the bad case:

grp->add_option_function<std::string>(
           "--file_path",
           [&](std::string_view str) {
               //do something here
           },
           "File Path")
        ->transform(MutatingValidator())
        ->check(NonMutatingValidator());

// where MutatingValidator is defined as follows:

struct MutatingValidator : public Validator {
    MutatingValidator() {
        name_ = "MutatingValidator";
        func_ = [](const std::string &str) {
            return std::string("My Sample mutating validator");
        };
    }
};

I also posted on stackoverflow:
https://stackoverflow.com/questions/70104056/cli11-check-never-called-after-transform-validator

**Comment by phlptp** [comment, marked as answer]
For a validator the return string is an error message.  Any changes to the string should be modified in place.  

So since your function returns a string it will be interpreted as an error and cease processing at that point.  so the check would never get called.  

The signature of the function call for a mutation validator is `std::string(std::string &)`
The input is the argument to process (passed by reference so it can be modified) the output is an error string if any.

**Reply by dhsu123** [reply]
Thanks Philip for the quick response. This worked. i was confused by this statement where it talks about throwing a ValidationError when there is an error vs just returning a non empty string:

https://cliutils.github.io/CLI11/book/chapters/validators.html:
"A transform validator comes in one form, a function with the signature std::string(std::string). The function will take a string and return the modified version of the string. If there is an error, the function should throw a CLI::ValidationError with the appropriate reason as a message."

**Reply by phlptp** [reply]
That chapter probably needs an update then.  Created an issue to address it.
